packages/cli/cli/src/commands/upgrade/upgradeGenerator.ts

```diff
@@ -1,30 +1,104 @@
+    const generatorGroups = parsedDocument.get("groups");
+    if (generatorGroups == null) {
+        context.failAndThrow("No groups were found within the generators configuration, no generators to upgrade.");
+        return undefined;
+    }
+    if (!YAML.isMap(generatorGroups)) {
+        context.failAndThrow(`Expected 'groups' to be a map in ${path.relative(process.cwd(), filepath)}`);
+        return undefined;
+    }
+    context.logger.info(`Groups found: ${generatorGroups.toString()}`);
+
+    for (const groupBlock of generatorGroups.items) {
+        const groupName = groupBlock.key;
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-arguments
+        const group = groupBlock.value as YAML.YAMLMap<string, YAML.YAMLSeq<YAML.YAMLMap<unknown, unknown>>>;
+        if (!YAML.isMap(group)) {
+            context.failAndThrow(
+                `Expected group ${groupName} to be a map in ${path.relative(process.cwd(), filepath)}`
+            );
+            continue;
+        }
+
+        if (groupFilter != null && group.get("name") !== groupName) {
+            continue;
+        }
+
+        const generators = group.get("generators");
+
+        if (!YAML.isSeq(generators)) {
+            context.failAndThrow(
+                `Expected group ${groupName} to have a 'generators' key in ${path.relative(process.cwd(), filepath)}`
+            );
+            continue;
+        }
+        context.logger.info(`Generators found: ${generators.toString()}`);
+        for (const generator of generators.items) {
+            if (!YAML.isMap(generator)) {
+                context.failAndThrow(
+                    `Expected generator in group ${groupName} to be a map in ${path.relative(process.cwd(), filepath)}`
+                );
+            }
+            if (generatorFilter != null && generator.get("name") !== generatorFilter) {
+                continue;
+            }
+
+            const generatorName = generator.get("name") as string;
+            const normalizedGeneratorName = generatorsYml.getGeneratorNameOrThrow(generatorName, context);
+
+            const currentVersion = generator.get("version") as string;
+
+            const latestVersion = await generatorsYml.getLatestGeneratorVersion(normalizedGeneratorName, context);
+            context.logger.info(`${generatorName}, ${currentVersion}, ${latestVersion}`);
+
+            if (latestVersion == null) {
+                continue;
+            }
+
+            // check if the versions have the same major version, if not fail, if yes do the upgrade
+            const currentMajor = semver.major(currentVersion);
+            const latestMajor = semver.major(latestVersion);
+            if (currentMajor === latestMajor) {
+                generator.set("version", latestVersion);
+                context.logger.info(chalk.green(`${generatorName} has been upgraded to latest in group: ${groupName}`));
+            }
```

Comment:
should we try to upgrade to the latest patch/minor?

Reply:
I'm not sure how we'd get it, do you have thoughts?
Basically right now we add a "latest" tag to any non-RC to be able to get the latest version
But to get the latest patch/minor we'd need to pull all the images, then grab the version tag, then parse and sort them, which seems rough?